be/src/runtime/runtime_filter_cache.cpp

```diff
@@ -0,0 +1,217 @@
+        : _num_slots(1L << log2_num_slots),
+          _slot_mask(_num_slots - 1),
+          _mutexes(_num_slots),
+          _filter_maps(_num_slots),
+          _event_mutexes(_num_slots),
+          _event_maps(_num_slots) {}
+RuntimeFilterCache::~RuntimeFilterCache() {
+    if (_clean_thread) {
+        this->stop_clean_thread();
+        _clean_thread->join();
+    }
+}
+Status RuntimeFilterCache::init() {
+    try {
+        _clean_thread = std::make_shared<std::thread>(_clean_thread_func, this);
+        return Status::OK();
+    } catch (...) {
+        return Status::InternalError("Fail to create clean_thread of RuntimeFilterCache");
+    }
+}
+void RuntimeFilterCache::_clean_thread_func(RuntimeFilterCache* cache) {
+    Thread::set_thread_name(cache->clean_thread(), "rf_cache_clr");
+    while (!cache->is_stopped()) {
+        cache->_clean_filters();
+        cache->_clean_events();
+        std::this_thread::sleep_for(milliseconds(100));
+    }
+}
+void RuntimeFilterCache::_clean_events() {
+    for (auto i = 0; i < _num_slots; ++i) {
+        auto& mutex = _event_mutexes[i];
+        auto& map = _event_maps[i];
+        std::unique_lock write_lock(mutex);
+        for (auto it = map.begin(); it != map.end();) {
+            if (it->second->is_expired()) {
+                it = map.erase(it);
+            } else {
+                ++it;
+            }
+        }
+    }
+}
+void RuntimeFilterCache::_clean_filters() {
+    for (auto i = 0; i < _num_slots; ++i) {
+        auto& mutex = _mutexes[i];
+        auto& map = _filter_maps[i];
+        std::unique_lock write_lock(mutex);
+        for (auto it = map.begin(); it != map.end();) {
+            if (it->second->is_expired()) {
+                it = map.erase(it);
+            } else {
+                ++it;
+            }
+        }
+    }
+}
+void RuntimeFilterCache::put_if_absent(const TUniqueId& query_id, int filter_id, const JoinRuntimeFilterPtr& filter) {
+    const auto slot_idx = std::hash<size_t>()(query_id.lo) & _slot_mask;
+    auto& mutex = _mutexes[slot_idx];
+    auto& map = _filter_maps[slot_idx];
+    _cache_times.fetch_add(1);
+    std::unique_lock write_lock(mutex);
+    if (!map.count(query_id)) {
+        map[query_id] = std::make_shared<RfCacheValue>();
+    }
+    auto& cached = map[query_id];
+    cached->put_if_absent(filter_id, filter);
+}
+
+JoinRuntimeFilterPtr RuntimeFilterCache::get(const TUniqueId& query_id, int filter_id) {
+    const auto slot_idx = std::hash<size_t>()(query_id.lo) & _slot_mask;
+    auto& mutex = _mutexes[slot_idx];
+    auto& map = _filter_maps[slot_idx];
+    std::shared_lock read_lock(mutex);
+    auto it = map.find(query_id);
+    if (it == map.end()) {
+        return nullptr;
+    } else {
+        auto filter = it->second->get(filter_id);
+        _use_times.fetch_add(filter != nullptr);
+        return filter;
+    }
+}
+
+void RuntimeFilterCache::add_rf_event(const TUniqueId& query_id, int filter_id, std::string&& msg) {
```

Comment:
Should we return directly when `_enable_trace` is false?

Reply:
yes